File: 124-binary-tree-maximum-path-sum/124-binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.ans = float('-inf')
        def dfs(node):
            if not node:
                return 0
            leftSum = max(dfs(node.left), 0)
            rightSum = max(dfs(node.right), 0)
            
            path = max(node.val, leftSum + rightSum + node.val)
            
            self.ans = max(path, self.ans)
            return max(leftSum, rightSum) + node.val
        
        dfs(root)
        return self.ans
        
        

File: 124-binary-tree-maximum-path-sum/124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int ans = INT_MIN;
public:
    int maxPathSum(TreeNode* root) {
        recurse(root);
        return ans;
    }
    
    
    int recurse(TreeNode* node){
        if (node == NULL) {
            return 0;
        }
        
        int leftSum = max(recurse(node->left), 0);
        int rightSum = max(recurse(node->right), 0);
        
        int path = leftSum + rightSum + node->val;
        ans = max(ans, path);
        
        return max(leftSum, rightSum) + node->val;
    }
};